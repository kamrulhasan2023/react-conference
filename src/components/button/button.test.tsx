import { render, fireEvent } from '@testing-library/react';
import Button from '.';
import {
  buttonProps,
  outlineButtonProps,
  redirectButtonProps,
} from './mock-data';
import {
  buttonColorSchemes,
  buttonSizes,
  outlineButtonColorSchemes,
} from './interface';

it('renders button correctly', () => {
  const { getByRole } = render(<Button {...buttonProps} />);
  const element = getByRole('button');
  expect(element).toBeInTheDocument();
  expect(element).toMatchSnapshot();
});

it('renders redirect button correctly', () => {
  const { getByRole } = render(<Button.Redirect {...redirectButtonProps} />);
  const element = getByRole('link');
  expect(element).toBeInTheDocument();
  expect(element).toMatchSnapshot();
});

describe('Given a mock function is passed to button', () => {
  it('Expect mockFunction to be called on button click', () => {
    const mockFunction = jest.fn();

    const props = { ...buttonProps, clickHandler: mockFunction };
    const { getByRole } = render(<Button {...props} />);
    const button = getByRole('button');
    fireEvent.click(button);
    expect(mockFunction).toHaveBeenCalled();
  });
});

describe.each([true, false])(
  'Given open new tab configurations',
  (isOpenNewTab) => {
    it(`Expect redirect button to contain href and ${
      isOpenNewTab ? 'target=_blank' : 'target=_self'
    }`, () => {
      const props = {
        ...buttonProps,
        href: 'https://www.google.com/',
        isOpenNewTab: isOpenNewTab,
      };
      const { getByRole } = render(<Button.Redirect {...props} />);
      const link = getByRole('link');

      expect(link).toHaveAttribute('href', 'https://www.google.com/');

      if (isOpenNewTab) {
        expect(link).toHaveAttribute('target', '_blank');
      } else {
        expect(link).toHaveAttribute('target', '_self');
      }
    });
  }
);

describe.each(buttonSizes)('Given button size: %s', (size) => {
  it(`Expect button to have class ${size}`, () => {
    const props = { ...buttonProps, size: size };
    const { getByRole } = render(<Button {...props} />);
    const button = getByRole('button');
    expect(button.children[0]).toHaveClass(size);
  });
});

describe.each(buttonColorSchemes)('Given color scheme %s', (colorScheme) => {
  it(`Expect button to have class ${colorScheme}`, () => {
    const props = { ...buttonProps, colorScheme: colorScheme };
    const { getByRole } = render(<Button {...props} />);
    const button = getByRole('button');

    expect(button.children[0]).toHaveClass(colorScheme);
  });
});

describe.each(outlineButtonColorSchemes)(
  'Given color scheme %s',
  (colorScheme) => {
    it(`Expect outline button to have class ${colorScheme}`, () => {
      const props = { ...outlineButtonProps, colorScheme: colorScheme };
      const { getByRole } = render(<Button {...props} />);
      const button = getByRole('button');

      expect(button.children[0]).toHaveClass(colorScheme);
    });
  }
);

describe.each([true, false])('Given iconIsLeft boolean', (iconIsLeft) => {
  it(`Expect button to ${
    iconIsLeft ? 'not' : ''
  } contain class: flex-row-reverse`, () => {
    const props = { ...buttonProps, iconIsLeft: iconIsLeft };
    const { getByRole } = render(<Button {...props} iconName="Arrow Right" />);
    const button = getByRole('button');

    if (iconIsLeft) {
      expect(button.children[0]).toHaveClass('flex-row-reverse');
    } else {
      expect(button.children[0]).not.toHaveClass('flex-row-reverse');
    }
  });
});
