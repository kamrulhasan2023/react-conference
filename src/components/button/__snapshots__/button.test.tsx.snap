// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders button correctly 1`] = `
<button
  class="inline-flex focus:outline-none focus-visible:outline-none focus:shadow-none focus-visible:shadow-none"
  type="button"
>
  <span
    aria-label="Button"
    class="relative z-10 group/button overflow-hidden inline-flex items-center justify-center gap-3 cursor-pointer text-white hover:text-white leading-[1.1] button bg-primary-text-white hover-bg-success-text-white lg"
  >
    <span>
      Button
    </span>
    <span
      class="absolute z-[-1] inset-0 invisible opacity-0 group-hover/button:visible group-hover/button:opacity-100 button-bg"
    />
  </span>
</button>
`;

exports[`renders redirect button correctly 1`] = `
<a
  class="inline-flex focus:outline-none focus-visible:outline-none focus:shadow-none focus-visible:shadow-none"
  href="https://www.google.com/"
  target="_blank"
>
  <span
    aria-label="Redirect Button"
    class="relative z-10 group/button overflow-hidden inline-flex items-center justify-center gap-3 cursor-pointer text-white hover:text-white leading-[1.1] button bg-primary-text-white hover-bg-success-text-white lg"
  >
    <span>
      Redirect Button
    </span>
    <span
      class="absolute z-[-1] inset-0 invisible opacity-0 group-hover/button:visible group-hover/button:opacity-100 button-bg"
    />
  </span>
</a>
`;
